include/thread_pool/slotted_bag.hpp

```diff
@@ -0,0 +1,169 @@
+
+        Slot(const Slot&) = delete;
+        Slot& operator=(const Slot&) = delete;
+
+        Slot(Slot&& rhs) = default;
+        Slot& operator=(Slot&& rhs) = default;
+    };
+
+public:
+
+    /**
+    * @brief SlottedBag Constructor.
+    * @param size Power of 2 number - queue length.
+    * @throws std::invalid_argument if size is not a power of 2.
+    */
+    explicit SlottedBag(size_t size);
+
+    SlottedBag(SlottedBag& rhs) = delete;
+    SlottedBag& operator=(SlottedBag& rhs) = delete;
+
+    SlottedBag(SlottedBag&& rhs) = default;
+    SlottedBag& operator=(SlottedBag&& rhs) = default;
+
+    /**
+    * @brief fill Fill the specified slot.
+    * @param id The id of the slot to fill.
+    * @throws std::runtime_error if the slot is already full.
+    * @note Other exceptions may be thrown if the single-producer-per-slot
+    * semantics are violated.
+    */
+    void fill(size_t id);
+
+    /**
+    * @brief empty Empties the slot with the specified id.
+    * @param id The id of the slot to empty.
+    * @returns true if the state of the bag was changed, false otherwise. 
+    * @note The slot will be empty after the call regardless of the return value.
+    */
+    bool empty(size_t id);
+
+    /**
+    * @brief tryEmptyAny Try to empty any slot in the bag.
+    * @param id The emptied slot id will be stored in this variable upon success.
+    * @return true upon success, false otherwise.
+    * @note Other exceptions may be thrown if the single-producer-per-slot
+    * semantics are violated.
+    */
+    bool tryEmptyAny(size_t& id);
+
+private:
+    Queue<Slot*> m_queue;
+    std::vector<Slot> m_slots;
+};
+
+template <template<typename> class Queue>
+inline SlottedBag<Queue>::SlottedBag(size_t size)
+    : m_queue(size)
+    , m_slots(size)
+{
+    for (auto i = 0u; i < size; i++)
+        m_slots[i].id = i;
+}
+
+template <template<typename> class Queue>
+inline void SlottedBag<Queue>::fill(size_t id)
+{
+    switch (m_slots[id].state.exchange(Slot::State::QueuedValid, std::memory_order_acq_rel))
+    {
+    case Slot::State::NotQueued:
+        // No race, single producer per slot.
+        while (!m_queue.pushWeak(&m_slots[id])); // Queue should never overflow. Use weak form for performance.
+        break;
+
+    case Slot::State::QueuedValid:
+        throw std::runtime_error("The item has already been added to the bag.");
+
+    case Slot::State::QueuedInvalid:
+        // Item will still be in the queue. We are done.
+        break;
+    }
+}
+
+template <template<typename> class Queue>
+inline bool SlottedBag<Queue>::empty(size_t id)
+{
+    // This consumer action is solely responsible for an indiscriminant QueuedValid -> QueuedInvalid state transition.
+    auto state = Slot::State::QueuedValid;
+    return m_slots[id].state.compare_exchange_strong(state, Slot::State::QueuedInvalid, std::memory_order_acq_rel);
+}
+
+template <template<typename> class Queue>
+inline bool SlottedBag<Queue>::tryEmptyAny(size_t& id)
```

Comment:
There are two alternatives interfaces you can use for this function to avoid passing id by reference:
1) use std::optional if the compiler is C++17 compatible
2) return a std::tuple<bool, size_t>

Reply:
We're still using C++14, but I'll switch the signature over to the tuple form.

Reply:
Fixed in #3.